src/loaders/css-module-dts-loader/loader.ts

```diff
@@ -0,0 +1,125 @@
+import { createSourceFile, forEachChild, Node, ScriptTarget, SyntaxKind } from 'typescript';
+import { statSync } from 'fs';
+import { resolve, dirname } from 'path';
+const DtsCreator = require('typed-css-modules');
+const { getOptions } = require('loader-utils');
+const instances = require('ts-loader/dist/instances');
+
+type TSLoaderInstances = {
+	files: {
+		[key: string]: boolean;
+	}
+}
+
+type DtsResult = {
+	writeFile(): Promise<void>;
+}
+
+type DtsCreatorInstance = {
+	create(filePath: string, initialContents: boolean, clearCache: boolean): Promise<DtsResult>;
+}
+
+type Webpack = {
+	resourcePath: string;
+	async(): (error: Error | null, result: string, sourceMap?: string) => void;
+}
+
+const creator: DtsCreatorInstance = new DtsCreator();
+
+const mTimeMap = new Map<string, Date>();
+
+function generateDTSFile(filePath: string): Promise<void> {
+	const { mtime } = statSync(filePath);
+	const lastMTime = mTimeMap.get(filePath);
+
+	if (!lastMTime || mtime > lastMTime) {
+		mTimeMap.set(filePath, mtime);
+		return creator.create(filePath, false, true).then((content) => {
+			return content.writeFile();
+		});
+	}
+	else {
+		return Promise.resolve();
+	}
+}
+
+function checkNodeForCSSImport(node: Node, instance?: TSLoaderInstances): Promise<void> {
+	if (node.kind === SyntaxKind.StringLiteral) {
+		const importPath = node.getText().replace(/\'|\"/g, '');
+		if (/.css$/.test(importPath)) {
+			const parentFileName = node.getSourceFile().fileName;
+
+			if (instance) {
+				instance.files[parentFileName] = false;
+			}
+
+			const absoluteCssFileName = resolve(dirname(parentFileName), importPath);
+			return generateDTSFile(absoluteCssFileName);
+		}
+		else {
+			return Promise.resolve();
+		}
+	}
+	else {
+		return Promise.resolve();
+	}
+}
+
+function checkNode(node: Node, instance?: TSLoaderInstances): Promise<any> {
```

Comment:
let me clarify, I think it would be easier to understand if you performed all the sync traversal of the ast first, then operate on the results of that which could then be async